jwks_test.go

```diff
@@ -141,47 +178,157 @@ func TestJWKs(t *testing.T) {
+	token3 := "eyJraWQiOiIxZjEyOGFkZSIsInR5cCI6IkpXVCIsImFsZyI6IkVTMjU2In0.eyJzdWIiOiJSZWJlY2NhIiwiYXVkIjoiQWxpY2UiLCJpc3MiOiJqd2tzLXNlcnZpY2UuYXBwc3BvdC5jb20iLCJleHAiOjE2MjQ3NTkzODIsImlhdCI6MTYyNDc1OTM3NywianRpIjoiMzU2MWY4MDctNDRkNi00OWE5LWFlYWItMmQ1MjQ2YWYxNDhlIn0.5eZbJlvnaFsRwPhBHmXljp9vgsrB0Q9d3dSz4va29ahTKsFGFo8tYy0e69ehqSb-dbFy9azRRtygwwtYuaEFuA"
+	token4 := "eyJraWQiOiIyZDQ3NjUwYSIsInR5cCI6IkpXVCIsImFsZyI6IlBTMzg0In0.eyJzdWIiOiJGcmVkYSIsImF1ZCI6Ikx1Y2lhIiwiaXNzIjoiandrcy1zZXJ2aWNlLmFwcHNwb3QuY29tIiwiZXhwIjoxNjI0ODA0MTk0LCJpYXQiOjE2MjQ4MDQxODksImp0aSI6IjdjNTQ2Y2RmLTYwMTEtNDI3Ny04Y2Q0LTMwNjZmZTYwNTExZSJ9.hQm-OP_MMk8_S13-ohiINRuDP2IlCiB3yn8Ov6qTjeFbq4gZ6MegeJH_qiZOvXqlzOAwpwd5P4nm5JeS6LlNGdW6V_agwYwnAd08GI7APQNRib692_sEk1DKdSk-S-Y8V_ZAgeTT8asdaSDw4EBPxkDvROcuEqesZrfqnrOcpdqqa2BcmwX8q5sLtQ8TMp4cOvEZg-J8_0j2kdCUkv_n9ZdsRoA3EUT8M1bYqnGRRxIRqflsm-S_xq3HxMAnPF5hPlqIKFVKuRsU0SKgcHZGwXpuK2lJqPobl6MI987tGrc9sPPFzVkNYxeltcxu34-ZjzN6iCQN8r0w-mfqCZav7A"
+
+	// Use the JWKs jwk.KeyFunc to parse the tokens signed with unknown kids at nearly the same time.
+	waitGroup := sync.WaitGroup{}
+	waitGroup.Add(3)
+	go func() {
+		defer waitGroup.Done()
+		if _, parseErr := jwt.Parse(token1, jwks.KeyFunc); parseErr != nil {
+			if errors.Is(parseErr, jwt.ErrInvalidKeyType) {
+				t.Errorf("Invaild key type selected.\nError:%s\n", parseErr.Error())
+			}
+		}
+	}()
+	go func() {
+		defer waitGroup.Done()
+		if _, parseErr := jwt.Parse(token2, jwks.KeyFunc); parseErr != nil {
+			if errors.Is(parseErr, jwt.ErrInvalidKeyType) {
+				t.Errorf("Invaild key type selected.\nError:%s\n", parseErr.Error())
+			}
+		}
+	}()
+	go func() {
+		defer waitGroup.Done()
+		if _, parseErr := jwt.Parse(token3, jwks.KeyFunc); parseErr != nil {
+			if errors.Is(parseErr, jwt.ErrInvalidKeyType) {
+				t.Errorf("Invaild key type selected.\nError:%s\n", parseErr.Error())
+			}
+		}
+	}()
+	if _, parseErr := jwt.Parse(token4, jwks.KeyFunc); parseErr != nil {
+		if errors.Is(parseErr, jwt.ErrInvalidKeyType) {
+			t.Errorf("Invaild key type selected.\nError:%s\n", parseErr.Error())
+			t.FailNow()
+		}
+	}
+	waitGroup.Wait()
+
+	// Confirm the JWKs was only refreshed once. (Refresh counter was first incremented on the creation of the JWKs.)
+	refreshMux.Lock()
+	expected := uint(2)
+	if refreshes != expected {
+		t.Errorf("An incorrect number of refreshes occurred.\n  Expected: %d\n  Got: %d\n", expected, refreshes)
+		t.FailNow()
+	}
+	refreshMux.Unlock()
+
+	// Wait for the rate limiter to take the next queue.
+	time.Sleep(refreshRateLimit + time.Millisecond*100)
+	refreshMux.Lock()
+	expected = uint(3)
+	if refreshes != expected {
+		t.Errorf("An incorrect number of refreshes occurred.\n  Expected: %d\n  Got: %d\n", expected, refreshes)
+		t.FailNow()
+	}
+	refreshMux.Unlock()
+
+	// Wait for the refresh interval to occur.
+	time.Sleep(refreshInterval + time.Millisecond*100)
```

Comment:
Another thing I'm not fond of in this PR are the sleep statements that my brain thinks are required to force some testing paths.

If there's a way around these, I'd like to consider that as well.